lib/zstd.h

```diff
@@ -2674,6 +2690,120 @@ ZSTDLIB_STATIC_API size_t ZSTD_decompressBlock(ZSTD_DCtx* dctx, void* dst, size_
+ *     windowSize if a dictionary is present, see doc/zstd_compression_format.md
+ *     for details.
+ *
+ * The user-provided function shall return a size_t representing the number of
+ * sequences written to outSeqs. This return value will be treated as an error
+ * code if it is greater than outSeqsCapacity. The return value must be non-zero
+ * if srcSize is non-zero. The ZSTD_EXTERNAL_MATCHFINDER_ERROR macro is provided
+ * for convenience, but any value greater than outSeqsCapacity will be treated as
+ * an error code.
+ *
+ * If the user-provided function does not return an error code, the sequences
+ * written to outSeqs must be a valid parse of the src buffer. Data corruption may
+ * occur if the parse is not valid. A parse is defined to be valid if the
+ * following conditions hold:
+ *   - The sum of matchLengths and literalLengths is equal to srcSize.
+ *   - All sequences in the parse have matchLength != 0, except for the final
+ *     sequence. matchLength is not constrained for the final sequence.
+ *   - All offsets respect the windowSize parameter as specified in
+ *     doc/zstd_compression_format.md.
+ *
+ * zstd will only validate these conditions (and fail compression if they do not
+ * hold) if the ZSTD_c_validateSequences cParam is enabled. Note that sequence
+ * validation has a performance cost.
+ *
+ * If the user-provided function returns an error, zstd will either fall back
+ * to an internal matchfinder or fail the compression operation. The user can
+ * choose between the two behaviors by setting the
+ * ZSTD_c_enableMatchFinderFallback cParam. Fallback compression will follow any
+ * other cParam settings, such as compression level, the same as in a normal
+ * compression operation.
+ *
+ * The user shall instruct zstd to use a particular ZSTD_externalMatchFinder_F
+ * function by calling ZSTD_registerExternalMatchFinder(cctx, externalMatchState,
+ * externalMatchFinder). This setting will persist until the next parameter reset
+ * of the CCtx.
+ *
+ * The externalMatchState must be initialized by the user before calling
+ * ZSTD_registerExternalMatchFinder. The user is responsible for destroying the
+ * externalMatchState.
+ *
+ * LIMITATIONS
+ * This API is currently incompatible with long-distance matching. As mentioned
+ * above, history buffers (stream history, dictionaries) are currently not
+ * supported. We plan to remove these limitations in the future.
+ */
+
+#define ZSTD_EXTERNAL_MATCHFINDER_ERROR ((size_t)(-1))
+
+typedef size_t ZSTD_externalMatchFinder_F (
+  void* externalMatchState,
+  ZSTD_Sequence* outSeqs, size_t outSeqsCapacity,
+  const void* src, size_t srcSize,
+  const void* dict, size_t dictSize,
+  int compressionLevel,
+  size_t windowSize
+);
+
+/*! ZSTD_registerExternalMatchFinder() :
+ * Instruct zstd to use an external matchfinder function.
+ *
+ * The externalMatchState must be initialized by the caller, and the caller is
```

Comment:
There are several compression APIs available to users. Could you address which APIs are supported and which are not?

Reply:
Thanks for your question! After writing the answer below, I realized that the "Limitations" section of the API docs can use some additional details. I will update the docs along the lines of the answer below.

Please let me know if you have any other questions that are not answered by the docs. It's super helpful to have your feedback in this area since it's impossible for me to look at the docs with fresh eyes.

I am probably going to land the PR today (after addressing your question in the docs), but feel free to continue leaving comments after it lands. We can always land a second PR to update the docs if we need to.

---

All compression APIs are supported. There are only two limitations:

* The `ZSTD_c_enableLongDistanceMatching` cParam is not supported. **Compression will fail if it is enabled and the user tries to compress with an external matchfinder.**
    * Note that `enableLongDistanceMatching` is enabled by default in some cases (see its documentation for details). The user needs to be careful not to use the external matchfinder in those cases.

* History buffers are not supported. Concretely, zstd will always pass `dictSize == 0` to the external matchfinder (for now). This has two implications:
    * Dictionaries are not supported. Compression will **not** fail if the user references a dictionary, but the dictionary won't have any effect.
    * Stream history is not supported. All compression APIs, including streaming APIs, work with the external matchfinder, but the external matchfinder won't get any history from the previous block.

Long-term, we plan to overcome both limitations. There's no technical reason why we can't. It's purely a question of engineering effort.

Reply:
Thanks, I'll do some test. If there are any questions, we can communicate the details by email.